### PR TITLE
feat:core: add RedeemHook logic

### PR DESCRIPTION
**Pull request by ramenforbreakfast**
So these tests are similar to David's where since `FeeRebates` require actually using `Collateral` and `BaseToken` for transferring tokens, and the caller is restricted to that same `Collateral` contract being used in the logic, we require David's `getSignerForCollateral` workaround.

I chose `FakeContract<Contract>` for `Collateral` to ensure no `Collateral` logic is actually simulated, which would make setup much more complicated. Since there was no existing `FakeContract` fixture for `Collateral`, I have added one.

### CODE REVIEW

File: apps/smart-contracts/core/contracts/RedeemHook.sol [hunk 1]
```diff
@@ -18,7 +19,15 @@ contract RedeemHook is
     address sender,
     uint256 amountBeforeFee,
     uint256 amountAfterFee
-  ) external virtual override onlyAllowedCallers {}
+  ) external virtual override onlyAllowedCallers {
+    require(_allowlist.isIncluded(sender), "sender not allowed");
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/RedeemHook.sol:23]: nit:
```suggestion
    require(_allowlist.isIncluded(sender), "redeemer not allowed");
```

File: apps/smart-contracts/core/test/RedeemHook.test.ts [hunk 2]
```diff
@@ -102,4 +111,75 @@ describe('=> RedeemHook', () => {
       await redeemHook.connect(deployer).setTokenSender(tokenSender.address)
     })
   })
+
+  describe('hook', () => {
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/RedeemHook.test.ts:115]: nit: (same for others)
```suggestion
  describe('# hook', () => {
```

File: apps/smart-contracts/core/contracts/RedeemHook.sol [hunk 3]
```diff
@@ -18,7 +19,15 @@ contract RedeemHook is
...
+    uint256 fee = amountBeforeFee - amountAfterFee;
+    if (fee > 0) {
+      IERC20 collateral = IPrePOMarket(msg.sender).getCollateral();
+      collateral.transferFrom(msg.sender, _treasury, fee);
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/contracts/RedeemHook.sol:27]: nit:
```suggestion
      IPrePOMarket(msg.sender).getCollateral().transferFrom(msg.sender, _treasury, fee);
```

File: apps/smart-contracts/core/test/RedeemHook.test.ts [hunk 4]
```diff
@@ -102,4 +111,75 @@ describe('=> RedeemHook', () => {
...
+            ceilingLongPayout: parseEther('0.8'),
+            floorValuation: BigNumber.from(10000000),
+            ceilingValuation: BigNumber.from(20000000),
+            expiryTime: nowPlusMonths(2),
+          },
+          user.address,
+          treasury.address
+        )
+        marketSigner = await getSignerForContract(market)
+        await redeemHook.setAllowedCallers([market.address], [true])
```
Comment (XavierEkkel) [inline apps/smart-contracts/core/test/RedeemHook.test.ts:155]: for these tests why is this necessary? couldn't you just set `user` as the allowed caller?

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

File: apps/smart-contracts/core/test/RedeemHook.test.ts (hunk 4 above)
Comment (ramenforbreakfast) [reply apps/smart-contracts/core/test/RedeemHook.test.ts:155]: No, because we wrap `msg.sender` with `IPrePOMarket` to grab the `Collateral` contract with `getCollateral`, this would not be possible with a normal signer. Thus we need to imitate the contract as a signer.

However, I have noticed that we do not even need to instantiate a smock market here, since all it needs is the collateral, so we can just use a `FakeContract` here and directly return `Collateral`.
